server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java

```diff
@@ -84,6 +86,19 @@ public DataStream rollover(Index newWriteIndex) {
         return new DataStream(name, timeStampField, backingIndices, generation + 1);
     }
 
+    /**
+     * Removes the specified backing index and returns a new {@code DataStream} instance with
+     * the remaining backing indices.
+     *
+     * @param index the backing index to remove
+     * @return new {@code DataStream} instance with the remaining backing indices
+     */
+    public DataStream removeBackingIndex(Index index) {
+        List<Index> backingIndices = new ArrayList<>(indices);
+        backingIndices.remove(index);
+        return new DataStream(name, timeStampField, backingIndices, generation);
```

Comment:
Would be good to assert that `backingIndices.size() == indices.size() - 1`?

Reply:
@henningandersen, I've added that assertion in https://github.com/elastic/elasticsearch/pull/55802